src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs

```diff
@@ -167,5 +174,58 @@ private static IReadOnlyList<PackageDependency> GetTransitivePackagesForLibrary(
 
             return null;
         }
+
+        internal static TransitivePackageReference MergeTransitiveOrigin(PackageReference currentPackage, TransitiveEntry transitiveEntry)
+        {
+            if (currentPackage == null)
+            {
+                throw new ArgumentNullException(nameof(currentPackage));
+            }
+            if (transitiveEntry == null)
+            {
+                throw new ArgumentNullException(nameof(transitiveEntry));
+            }
+
+            var transitiveOrigins = new SortedSet<PackageReference>(PackageReferenceMergeComparer);
+            transitiveEntry?.Keys?.ForEach(fwRuntimePair =>
+            {
+                if (fwRuntimePair != null)
+                {
+                    IList<PackageReference> fwTransitiveOrigins = transitiveEntry[fwRuntimePair];
+                    if (fwTransitiveOrigins != null)
+                    {
+                        foreach (PackageReference transitiveOrigin in fwTransitiveOrigins)
+                        {
+                            if (transitiveOrigin?.PackageIdentity != null && transitiveOrigin.PackageIdentity.Id != null && transitiveOrigin.PackageIdentity.Version != null)
```

Comment:
> What is the scenario where the transitiveEntry[fwRuntimePair] would have a null item and therefore fwTransitiveOrigins would contain a null PackageReference that we're calling transitiveOrigin?

This can occur if `installedPackages` contains a null `PackageReference` object:

https://github.com/NuGet/NuGet.Client/blob/d21029641e348e06742f5f563edf084ed811ec3a/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs#L270

> is this a valid scenario ...?

Probably not, because `GetInstalledAndTransitivePackagesAsync()` always return a non-null list, but there's no code that guarantee non-null PackageReference objects.

> if so, are we testing that this doesn't actually occur?

This is tested at `MergeTransitiveOrigin_WithNullTransitiveEntryList_ReturnsMergedTransitiveOrigins`